compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java

```diff
@@ -4577,6 +4582,45 @@ private BLangExpression createConditionForErrorFieldMatchPatterns(
         return statementExpression;
     }
 
+    private BType createMatchingRecordType(BLangErrorFieldMatchPatterns errorFieldMatchPatterns) {
+        BRecordTypeSymbol detailRecordTypeSymbol = new BRecordTypeSymbol(
+                SymTag.RECORD,
+                Flags.PUBLIC,
+                names.fromString("$anonErrorType$" + UNDERSCORE + recordCount++ + "$detailType"),
+                env.enclPkg.symbol.pkgID, null, null, errorFieldMatchPatterns.pos, VIRTUAL);
+
+        detailRecordTypeSymbol.initializerFunc = createRecordInitFunc();
+        detailRecordTypeSymbol.scope = new Scope(detailRecordTypeSymbol);
+        detailRecordTypeSymbol.scope.define(
+                names.fromString(detailRecordTypeSymbol.name.value + "." +
+                        detailRecordTypeSymbol.initializerFunc.funcName.value),
+                detailRecordTypeSymbol.initializerFunc.symbol);
+
+        BRecordType detailRecordType = new BRecordType(detailRecordTypeSymbol);
```

Comment:
This segment is duplicated in several places (i.e `createDetailType()`), better if we can refactor that and move it out to a re-usable method.